### PR TITLE
Initial version of Dynamo capture for HigherOrderOperator

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #99483
* __->__ #99988
* #99987

This PR introduces a `wrap(body_fn, *args)` higher order operator
The semantics of `wrap(body_fn, *args)` is to just run `body_fn(*args)`

Underneath Dynamo, this PR makes it so that we rewrite calls to
`wrap(body_fn, *args)` with `wrap(new_fn, *new_args)` where `new_fn` has
no free variables. This PR does not update cond/map to use the new
mechanism yet (we do not support nn.Modues yet, will come in the future).

The design we take is:
- OutputGraph represents the graph being built by Dynamo that may be
compiled and executed.
- OutputGraph owns a root SubgraphTracer, where it builds the FX graph.
- OutputGraph may own multiple nested SubgraphTracers.
- When we need to trace the body function of a HigherOrderOperator, we
construct a new SubgraphTracer to build the graph of the body function.

Mechanically, when Dynamo sees a new `wrap` HigherOrderOperator with a
body function, it:
- Creates a new SubgraphTracer via OutputGraph.new_subtracer
- Executes the body function
This captures the body function into the graph on the new
SubgraphTracer while modifying the state of the OutputGraph. For
example, the OutputGraph may receive new GraphArgs, new guards, and new
side effects.

If capture of the body function fails, then Dynamo graph breaks on the
HigherOrderOperator.

Test Plan:
- added test/dynamo/test_higher_order_ops.py

Future:
- We're not actually able to tell Dynamo to completely graph break on the
HigherOrderOperator. Instead, when we do graph break, Dynamo begins
introspecting `HigherOrderOperator.__call__`. It should probably not do
this.
- Ideally we would error out on new SideEffects. I don't know how to do
this yet.
- We don't support dealing with nn.Modules yet (e.g. calling nn.Modules
or accessing attributes of tracked nn.Modules from a body_fn). There's
an open question on what should actually happen here
- Ideally we would rewrite map/cond to use the new mechanism but we need
to fix the previous bullet point before we can get there.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: test/dynamo/test_higher_order_ops.py [hunk 1]
```diff
@@ -0,0 +1,351 @@
+# Owner(s): ["module: dynamo"]
```
Comment (voznesenskym) [inline test/dynamo/test_higher_order_ops.py:1]: Can you add this to test/dynamo/test_dynamic_shapes.py please?

File: test/dynamo/test_higher_order_ops.py [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+
+        def f(x):
+            return wrap(lambda x: torch.sin(x), x)
+
+        x = torch.randn(3)
+        expected = f(x)
+        result = torch.compile(f, backend=mock)(x)
+
+        self.assertEqual(result, expected)
+        self.assertEqual(len(mock.graphs), 1)
```
Comment (voznesenskym) [inline test/dynamo/test_higher_order_ops.py:37]: Use compilecounter for this :P

```
counter = CompileCounter()
compiled = torch.compile(func, backend=counter)
self.assertEqual(counter.frame_count, 2)
```

Comment (voznesenskym) [reply test/dynamo/test_higher_order_ops.py:37]: frame_count should suffice for both counting the graphs and graph breaks, remove counters too.

File: test/dynamo/test_higher_order_ops.py [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+        def f(x):
+            return wrap(lambda x: x + global_var, x)
+
+        x = torch.randn(3)
+        expected = f(x)
+        result = torch.compile(f, backend=mock)(x)
+
+        self.assertEqual(result, expected)
+        self.assertEqual(len(mock.graphs), 1)
+        # wrap(fn, x, global_var)
```
Comment (voznesenskym) [inline test/dynamo/test_higher_order_ops.py:53]: rm comment

File: test/dynamo/test_higher_order_ops.py [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+        self.assertIsNotNone(re.search(r"wrap\(\w+, \w+, \w+\);", mock.graphs[0].code))
+        self.assertEqual(len(counters["graph_break"]), 0)
```
Comment (voznesenskym) [inline test/dynamo/test_higher_order_ops.py:55]: Don't do this, instead, compile with nopython=true

File: test/dynamo/test_higher_order_ops.py [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+    def test_capture_untracked_global(self):
+        counters.clear()
+        mock = MockBackend()
+
+        def f(x):
+            return wrap(lambda x: x + global_var, x)
+
+        x = torch.randn(3)
+        expected = f(x)
+        result = torch.compile(f, backend=mock)(x)
```
Comment (voznesenskym) [inline test/dynamo/test_higher_order_ops.py:49]: IMO prefer testing `torch._dynamo.optimize("eager")` - that's the frontend. This invokes aot_autograd and inductor, which is a fuller test. Keep unit tests to a unit (graph capture, in this case). 

Both are still worth having, but maybe lower down, or alongside this.

File: test/dynamo/test_higher_order_ops.py [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+
+        self.assertEqual(result, expected)
+        self.assertEqual(len(mock.graphs), 1)
+        # wrap(fn, x, global_var)
+        self.assertIsNotNone(re.search(r"wrap\(\w+, \w+, \w+\);", mock.graphs[0].code))
```
Comment (voznesenskym) [inline test/dynamo/test_higher_order_ops.py:54]: This feels brittle - consider iterating the graph nodes and asserting that none have this in the node attributes?

File: test/dynamo/test_higher_order_ops.py (hunk 2 above)
Comment (voznesenskym) [reply test/dynamo/test_higher_order_ops.py:37]: You also get opcounts, which this test/ these tests would benefit from

File: test/dynamo/test_higher_order_ops.py [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+
+        @torch.compile(backend=mock)
+        def f(x, y):
+            return wrap(inner, x, y)
+
+        result = f(x, y)
+
+        self.assertEqual(result, x + y + x)
+        self.assertEqual(len(mock.graphs), 1)
+        gm = mock.graphs[0]
```
Comment (voznesenskym) [inline test/dynamo/test_higher_order_ops.py:156]: If you need to extract the graph, use export, or CompileCounterWithBackend

File: torch/_dynamo/allowed_functions.py [hunk 8]
```diff
@@ -168,6 +168,11 @@ def _find_torch_objects(module):
         torch_object_ids[id(module)] = module.__name__
         for name, obj in list(module.__dict__.items()):
             if id(obj) not in torch_object_ids:
+                # Don't handle HigherOrderOperator as builtin
+                import torch._ops
+
+                if isinstance(obj, torch._ops.HigherOrderOperator):
+                    continue
```
Comment (voznesenskym) [inline torch/_dynamo/allowed_functions.py:175]: Add more detail in a comment as to why this is the way it is, please.

File: torch/_dynamo/output_graph.py [hunk 9]
```diff
@@ -454,6 +465,10 @@ def register_attr_or_module(
         assert not isinstance(source, ParamBufferSource)
 
         if isinstance(target, torch.Tensor):
+            if len(self.tracers) > 1:
+                unimplemented(
+                    "accessing attribute of nn.Module inside HigherOrderOperator"
+                )
```
Comment (voznesenskym) [inline torch/_dynamo/output_graph.py:471]: 1) Why here?
2) You may have to let constant source through, or add a comment about constant source.

File: torch/_dynamo/output_graph.py [hunk 10]
```diff
@@ -941,6 +968,21 @@ def create_proxy(
...
+                if not isinstance(arg, torch.fx.Proxy):
+                    # Is a constant
+                    continue
+                if arg in self.seen_proxies:
+                    continue
+                if not hasattr(arg, "node"):
+                    continue
+                if arg.node.name in self.input_name_to_proxy:
+                    continue
+                self.lift_freevar_to_input(arg)
```
Comment (voznesenskym) [inline torch/_dynamo/output_graph.py:984]: Don't I want to do this even if I don't have a parent?

File: torch/_dynamo/output_graph.py [hunk 11]
```diff
@@ -1015,3 +1060,51 @@ def create_graph_input(self, name, type_expr=None, before=False):
...
+
+    def is_name_bound(self, name):
+        if name in self.input_name_to_proxy:
+            return True
+        for proxy in self.seen_proxies:
+            if proxy.node.name == name:
+                return True
+        return False
+
+    def lift_freevar_to_input(self, proxy):
```
Comment (voznesenskym) [inline torch/_dynamo/output_graph.py:1072]: As noted above - you only call this under `if self.parent is not None:` 

Either make the invariant that this is safe to call at any time (Like you have on L1075) or assert that self.parent is not None here and call it unequivocally after name bound checking.

File: torch/_dynamo/output_graph.py [hunk 12]
```diff
@@ -1015,3 +1060,51 @@ def create_graph_input(self, name, type_expr=None, before=False):
...
+        self.create_graph_input(proxy.node.name)
+        self.lifted_freevars.add(proxy)
+        if self.parent is not None and not self.parent.is_name_bound(proxy.node.name):
+            self.parent.lift_freevar_to_input(proxy)
+
+
+# NOTE: [HigherOrderOperator tracing design]
+# Ignoring HigherOrderOperators for a moment,
+# OutputGraph represents the graph being built by Dynamo that may be compiled
```
Comment (voznesenskym) [inline torch/_dynamo/output_graph.py:1081]: If you write a general note on it, please add that it's lifecycle is 1:1 w/ a frame being processed.

File: torch/_dynamo/variables/torch.py [hunk 13]
```diff
@@ -828,7 +828,7 @@ def get_comparable_state(state):
                 )
             )
 
-        def speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
+        def old_speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
```
Comment (voznesenskym) [inline torch/_dynamo/variables/torch.py:831]: Oh no what happened that these split?

File: torch/_dynamo/variables/torch.py [hunk 14]
```diff
@@ -911,6 +911,53 @@ def speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
...
+                            tracer.create_graph_input(a.as_proxy().node.name)
+                            args.append(a)
+                        else:
+                            # call_function() needs a TensorVariable, therefore we construct
+                            # one with inner graph proxy.
+                            assert isinstance(a, torch.Tensor)
+                            proxy = tracer.create_graph_input("arg")
+                            args.append(
+                                wrap_fx_proxy(tx=tx, proxy=proxy, example_value=a)
+                            )
```
Comment (voznesenskym) [inline torch/_dynamo/variables/torch.py:931]: nit: Move this assert to the top of the for loop, and give it a message about only supporting tensor or TensorVariable.

Also, you can potentially rewrite it a little.
```suggestion
            for a in sub_args:
                assert isinstance(a, (TensorVariable, torch.Tensor)), "Only Tensor/Tensor VariableTracker allowed here."
                
                if isinstance(a, torch.Tensor):
                    # call_function() needs a TensorVariable, therefore we construct
                    # one with inner graph proxy.
                    proxy = tracer.create_graph_input("arg")
                    a = wrap_fx_proxy(tx=tx, proxy=proxy, example_value=a)
                    assert isinstance(a, TensorVariable)
                else:
                    tracer.create_graph_input(a.as_proxy().node.name)

                args.append(a)
```

File: torch/_dynamo/output_graph.py [hunk 15]
```diff
@@ -1015,3 +1060,51 @@ def create_graph_input(self, name, type_expr=None, before=False):
...
+# and executed. It holds a root SubgraphTracer where the FX graph is built.
+#
+# HigherOrderOperators are operators that take functions (represented by
+# torch.fx.GraphModule) as their arguments. When Dynamo encounters a
```
Comment (BowenBao) [inline torch/_dynamo/output_graph.py:1085]: I'm trying to understand what `(represented by torch.fx.GraphModule)` implies. Does it mean the functions must be `torch.fx.GraphModule` to begin with? But that doesn't feel right given it is mentioned later that "dynamo introspect the function and capture into GraphModule".

File: test/dynamo/test_higher_order_ops.py [hunk 16]
```diff
@@ -0,0 +1,351 @@
...
+            return wrap(lambda x: wrap(lambda x: x + global_var, x), x)
+
+        x = torch.randn(3)
+        result = f(x)
+
+        self.assertEqual(result, x + global_var)
+        self.assertEqual(len(mock.graphs), 1)
+        gm = mock.graphs[0]
+        self.assertIsNotNone(re.search(r"wrap\(\w+, \w+, \w+\);", gm.code))
+        self.assertIsNotNone(re.search(r"wrap\(\w+, \w+, \w+\);", gm.cond_body_1.code))
```
Comment (BowenBao) [inline test/dynamo/test_higher_order_ops.py:72]: Just putting a note here in case anybody else is wondering why it's called `cond_xxx`. It appears back then "cond()" was the only user of `add_subgraph` so the word ended up being hardcoded there.

File: torch/_dynamo/output_graph.py (hunk 15 above)
Comment (Chillee) [reply torch/_dynamo/output_graph.py:1085]: Basically, a higher order operator is something like `higher_order_op(f, args)`, where `f` is an arbitrary user-provided function. The problem we're dealing with is - what if `f` depends on things like global variables or any other value that isn't directly passed in to `args`?

Well, we want to lift them to arguments. And luckily, this is exactly what Dynamo can do :) It can find all values that we depend upon (regardless of whether they're local variables, closed over variables, or globals), and lift them to be explicitly inputs to the fx.GraphModule.

So, for example, if you write
```
y = ...
def f(x):
    return x + y
 
higher_order_op(f, inp)
```
we will automatically rewrite that into (more or less)
```
def f_graph(x, y):
    return x + y
higher_order_op(f_graph, inp, y)
```

File: test/dynamo/test_higher_order_ops.py (hunk 16 above)
Comment (zou3519) [reply test/dynamo/test_higher_order_ops.py:72]: Yes, that's on the list to change at some point

File: torch/_dynamo/output_graph.py [hunk 17]
```diff
@@ -289,7 +291,7 @@ def real_value_cache(self):
         return self.current_tracer.real_value_cache
 
     def create_graph_input(self, *args, **kwargs):
-        return self.current_tracer.create_graph_input(*args, **kwargs)
+        return self.root_tracer.create_graph_input(*args, **kwargs)
```
Comment (anijain2305) [inline torch/_dynamo/output_graph.py:294]: Why this change? Each SubgraphTracer will create its own graph inputs, right?

Comment (zou3519) [reply torch/_dynamo/output_graph.py:294]: A motivating example is when we begin tracking a new Tensor variable inside the body of `wrap`:
```
y = torch.randn(3)

@torch.compile
def f(x):
   return wrap(lambda x: x + y, x)
```

At some point, while introspecting the body of `wrap`, Dynamo will call `tx.output.create_graph_input` because it sees `y` and `y` has not been tracked. What should happen in this case is:
1. Dynamo promotes `y` to being a graph input of the root graph
2 Dynamo inserts `y` as an input to `wrap`'s body function

The mechanism behind (1) is that we change `tx.output.create_graph_input` to mean `tx.output.root_tx.create_graph_input`. The mechanism behind (2) is the same for whenever the SubgraphTracer sees a tracked free-variable: SubgraphTracer will lift by calling `lift_freevar_to_input` in `SubgraphTracer.create_proxy`.

NB: If we want to actually create a graph input for the body_function, then we would directly call `tracer.create_graph_input`, which happens at some places in this PR.

I can add this as a comment into this PR and am taking suggestions for how to make the design more sane because I agree this is strange.

File: test/dynamo/test_higher_order_ops.py (hunk 2 above)
Comment (zou3519) [reply test/dynamo/test_higher_order_ops.py:37]: I also want to assert on what the graphs look like. I could modify CompileCounter to also keep track of the graphs?

File: torch/_dynamo/output_graph.py (hunk 9 above)
Comment (zou3519) [reply torch/_dynamo/output_graph.py:471]: > Why here?

I'm not sure where else to do it. I've considered doing it in SubgraphTracer.create_proxy when it sees a `get_attr` call, but we can't actually tell what the get_attr call is for (e.g. if it is for a Tensor, or for a body function (which is represented as a GraphModule)) without adding more infra.

> You may have to let constant source through, or add a comment about constant source.

Good point, let me play around with it.

File: test/dynamo/test_higher_order_ops.py (hunk 2 above)
Comment (zou3519) [reply test/dynamo/test_higher_order_ops.py:37]: Below you mentioned CompileCounterWithBackend, which I think does what I want.

File: torch/_dynamo/output_graph.py (hunk 10 above)
Comment (zou3519) [reply torch/_dynamo/output_graph.py:984]: It already gets taken care of if you don't have a parent. If Dynamo sees an untracked Tensor in LOAD_GLOBAL or LOAD_DEREF, it will begin tracking it, and as a part of that it will lift the freevar to the graph input by calling `tx.output.create_graph_input`.

Maybe I should rename `lift_freevar_to_input` to `lift_tracked_freevar_to_input`

File: torch/_dynamo/output_graph.py (hunk 15 above)
Comment (zou3519) [reply torch/_dynamo/output_graph.py:1085]: We've got a bit of a vocabulary problem that needs to be resolved at some point. It will probably like the following:
- there is a set of higher order operators that accept arbitrary callables as body functions
- there is a set of higher order operators that accept only sanitized FX graphs (in the form of GraphModules) as inputs. 

For example, you may have functorch.experimental.map(f, xs) that can accept arbitrary functions as inputs, but after Dynamo captures `f` into `f_graph_safe`, it should also rewrite the call to `map` into a call to `torch._ops.map_sanitized(f_graph_safe, xs)`.

We might even have a third category of higher order operators depending on how we decide to handle nn.Modules: nn.Modules could be in the body function or we could ban them from the body function.

File: torch/_dynamo/variables/torch.py (hunk 13 above)
Comment (zou3519) [reply torch/_dynamo/variables/torch.py:831]: I cannot actually rewrite Dynamo capture for map() and cond() to use the new `speculate_subgraph`. This is because we do not support capturing nn.Modules or nn.Module attributes yet. That situation is tricky because there is an open question of if the nn.Module should appear in the body function or if we should completely trace out the nn.Module so that it does not.

I'm leaving that as a follow-up, but let me know if you want it in this PR.

File: torch/_dynamo/variables/torch.py (hunk 14 above)
Comment (zou3519) [reply torch/_dynamo/variables/torch.py:931]: SGTM, I was copy-pasting the original `speculate_subgraph`

File: torch/_dynamo/variables/torch.py (hunk 13 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/torch.py:831]: Followup is fine, but let's add a comment :)

**Review by voznesenskym — APPROVED** [review]

File: torch/_dynamo/output_graph.py (hunk 10 above)
Comment (voznesenskym) [reply torch/_dynamo/output_graph.py:984]: ack - this is fine, but add a comment :P

File: torch/_dynamo/output_graph.py (hunk 17 above)
Comment (zou3519) [reply torch/_dynamo/output_graph.py:294]: I ended up just deleting OutputGraph.create_graph_input. Anyone who wants to use this needs to explicitly say tx.output.current_tracer.create_graph_input or tx.output.root_tracer.create_graph_input

File: torch/_dynamo/output_graph.py (hunk 9 above)
Comment (zou3519) [reply torch/_dynamo/output_graph.py:471]: @voznesenskym do you have an example of what it means to have constant source? I am not sure how to construct a repro that gets here

File: torch/_dynamo/variables/torch.py [hunk 18]
```diff
@@ -911,6 +911,53 @@ def speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
...
+                    return (
+                        output,
+                        graph,
+                        lifted_freevars,
+                    )
+
+            except torch._dynamo.exc.Unsupported as ex:
+                tx.output.graph = graph_checkpoint
+                tx.restore_graphstate(checkpoint)
+                raise
```
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/_dynamo/variables/torch.py:959]: This does give us two restores, is that a problem? If introspection of the body function raises Unsupported, then:
- we will restore_graphstate here
- we will restore_graphstate to the call to HigherOrderOperator

At any rate, I think we need to restore tx.output.graph because we may have mutated it if we saw any untracked free variables

**Comment by zou3519** [comment]
Phew, resolved most of the comments. This should be ready for another pass-through if you folks are interested

**Review by anijain2305 — APPROVED** [review]
Awesome work!

File: torch/_dynamo/variables/torch.py [hunk 19]
```diff
@@ -1149,6 +1208,34 @@ def speculate_branch(branch):
...
+                [
+                    *args[1:],
+                ],
+                graph_checkpoint,
+                checkpoint,
+            )
+
+            body_name = add_subgraph(
+                "body", torch.fx.GraphModule(tx.output.nn_modules, body_graph)
+            )
```
Comment (bdhirsh) [inline torch/_dynamo/variables/torch.py:1230]: (probably dumb) nit: it looks like `add_subgraph` hardcodes the name of the submodule to be `cond_{name}_{i}`. Is `add_subgraph` general infra meant for use by all higher order ops?

File: torch/_dynamo/output_graph.py [hunk 20]
```diff
@@ -953,6 +990,46 @@ def create_proxy(
...
+        # will see the freevar when it gets loaded by Python bytecode.
+        # E.g. for Python 3.11 the bytecodes that may do this are LOAD_DEREF or
+        # LOAD_GLOBAL.
+        # There, the InstructionTranslator asks Dynamo to begin tracking the
+        # freevar by building a new Variable.
+        # Building a new Variable automatically lifts the freevar to be an
+        # input of the root SubgraphTracer.
+        #
+        # After that happens, since the variable is now being tracked,
+        # we are now back to case 1.
```
Comment (bdhirsh) [inline torch/_dynamo/output_graph.py:1018]: Confirming my understanding - is this note effectively saying that:

(1) we will always be in case 1 by the time we get to this code.

(2) Any "free variable" we encounter is guaranteed to already have been lifted into a graph input of the root graph. The additional work we need to do is *only* that we need to lift this free variable into inputs (recursively) of every nested higher-order-op subgraph that we're currently inside of, but not the root graph.

**Comment by bdhirsh** [comment]
Really cool!

File: torch/_dynamo/output_graph.py (hunk 20 above)
Comment (zou3519) [reply torch/_dynamo/output_graph.py:1018]: Yes! I'll make that clearer
